### PR TITLE
feat(interop): Add Hyphanet compatibility gate

### PR DESCRIPTION
**Pull request by leumor**
PR-193 interop-gate

## Summary
- Add `tools/interop` harness, baseline env, README, and FCP transcript/summary diagnostics under `build/interop-smoke/`.
- Launch Cryptad and a verified/configured Hyphanet baseline, exchange peers, and validate FCP handshake, node info, peer listing, CHK/SSK/USK cross-node content flows, and restart recovery.
- Add request-local `ConsecutiveRNFsCountAsSuccess` FCP insert strictness so the gate can avoid false `PutSuccessful` results after route-not-found outcomes.
- Preserve legacy baseline override variables, redact private insert URIs in transcripts, and force restart refetches past local datastore/cache with payload validation.
- Add focused unit coverage for FCP insert option plumbing, directory inserts, strict RNF behavior, and value semantics.

## Validation
- `./gradlew spotlessApply`
- `./gradlew :test --tests '*ClientPutBaseTest' --tests '*ClientPutMessageTest' --tests '*ClientPutDiskDirMessageTest' --tests '*ClientPutComplexDirMessageTest' --tests '*ClientPutDirTest' --tests '*FcpInsertOptionsTest' --tests '*DefaultFcpInsertContextHandleTest' --tests '*SingleBlockInserterTest' --tests '*CoreFcpInsertRuntimeSupportTest'`
- `./gradlew test`
- `python3 -m py_compile tools/interop/interop_smoke.py`
- `python3 tools/interop/interop_smoke.py --self-test`
- `bash -n tools/interop/run-hyphanet-interop-smoke.sh tools/interop/hyphanet-baseline.env`
- `INTEROP_SKIP_BUILD=1 INTEROP_CACHE_DIR=build/interop-cache tools/interop/run-hyphanet-interop-smoke.sh`

## Notes
- Baseline configuration used for the live gate: checked-in Hyphanet 1506 baseline defaults, using the local `build/interop-cache` cache and existing `build/cryptad-dist` (`INTEROP_SKIP_BUILD=1`).
- Live interop completed all required flows successfully, including strict post-restart CHK/SSK/USK refetch validation.
- Deferred: multi-OS matrix, long-lived USK subscription soak, opennet baseline testing, UI/browser tests, AppHost catalog remote fetch tests, and performance benchmarking.

### CODE REVIEW

File: tools/interop/interop_smoke.py [hunk 1]
```diff
@@ -425,22 +591,18 @@ def download_asset() -> None:
...
 
 def materialize_hyphanet_java_symlinks(extract_root: Path) -> None:
     java_dir = extract_root / "usr" / "share" / "java"
     if not java_dir.is_dir():
         return
 
-    version = re.escape(load_env_required("HYPHANET_VERSION"))
-    build = re.escape(load_env_required("HYPHANET_BUILD"))
+    version = re.escape(os.environ.get("HYPHANET_VERSION", "0.7.5"))
+    build = re.escape(os.environ.get("HYPHANET_BASELINE_VERSION") or os.environ.get("HYPHANET_BUILD", ""))
```
[the remark this answers is not in the document]
Comment (leumor) [reply tools/interop/interop_smoke.py]: Fixed in fae15ea2ab.

The .deb jar symlink materialization now infers Hyphanet-style versioned jar suffixes from the extracted filenames instead of depending on HYPHANET_VERSION=0.7.5, so newer verified baseline .deb artifacts can still create the unversioned wrapper.conf classpath entries. I also added a self-test with synthetic 0.8.0+1600 jars and verified:

- python3 -m py_compile tools/interop/interop_smoke.py
- python3 tools/interop/interop_smoke.py --self-test
- git diff --check
- INTEROP_SKIP_BUILD=1 INTEROP_CACHE_DIR=build/interop-cache tools/interop/run-hyphanet-interop-smoke.sh

File: tools/interop/hyphanet-baseline.env [hunk 2]
```diff
@@ -1,6 +1,24 @@
...
-HYPHANET_RELEASE_URL=https://github.com/hyphanet/fred/releases/download/${HYPHANET_RELEASE_TAG}/${HYPHANET_DEB_ASSET}
+# Defaults for tools/interop/run-hyphanet-interop-smoke.sh.
+#
+# Environment variables set by the caller take precedence. The checked-in default uses a verified
+# Hyphanet 1506 Debian package because that release contains the baseline runtime jars and wrapper
+# config needed to launch the node reproducibly on Linux CI.
+
+HYPHANET_BASELINE_VERSION="${HYPHANET_BASELINE_VERSION:-1506}"
+HYPHANET_BASELINE_JAR="${HYPHANET_BASELINE_JAR:-}"
+HYPHANET_BASELINE_URL="${HYPHANET_BASELINE_URL:-${HYPHANET_RELEASE_URL:-https://github.com/hyphanet/fred/releases/download/build01506/freenet_0.7.5+1506-1_amd64.deb}}"
```
[the remark this answers is not in the document]
Comment (leumor) [reply tools/interop/hyphanet-baseline.env]: Fixed in ecf2fa1742.

`hyphanet-baseline.env` now initializes the legacy release aliases before deriving the new baseline URL/checksum, so old invocations that set `HYPHANET_RELEASE_TAG`, `HYPHANET_DEB_ASSET`, and `HYPHANET_DEB_SHA256` now select the intended baseline instead of falling back to build01506. Explicit `HYPHANET_BASELINE_URL` / `HYPHANET_BASELINE_SHA256` still take precedence, and `HYPHANET_RELEASE_URL` continues to work as a legacy direct URL override.

Validated:

- `bash -n tools/interop/hyphanet-baseline.env tools/interop/run-hyphanet-interop-smoke.sh`
- default env source check: baseline and legacy URL/checksum aliases match build01506
- legacy tag/asset/checksum source check: baseline URL derives from the legacy tag and asset
- legacy direct `HYPHANET_RELEASE_URL` source check
- explicit `HYPHANET_BASELINE_URL` / `HYPHANET_BASELINE_SHA256` source check
- `git diff --check`

I also re-ran the cached live interop smoke. It started the default verified 1506 `.deb` baseline and completed handshake, peer exchange, CHK, SSK, and USK smoke, then timed out in the existing restart-recovery post-restart USK refetch path. That timeout is not caused by this env resolution change; diagnostics were captured under `build/interop-smoke/`.

File: tools/interop/run-hyphanet-interop-smoke.sh [hunk 3]
```diff
@@ -1,38 +1,46 @@
...
+  echo "Missing baseline definition: $BASELINE_ENV" | tee "$ROOT_DIR/build/interop-smoke/shell-error.txt" >&2
   exit 1
 fi
 
-if [[ ! -f "$BASELINE_ENV" ]]; then
-  echo "Missing baseline definition: $BASELINE_ENV" >&2
-  exit 1
-fi
+OUT_DIR="${INTEROP_WORKDIR:-${INTEROP_OUT_DIR:-$ROOT_DIR/build/interop-smoke}}"
+CACHE_DIR="${INTEROP_CACHE_DIR:-$OUT_DIR/downloads}"
```
[the remark this answers is not in the document]
Comment (leumor) [reply tools/interop/run-hyphanet-interop-smoke.sh]: Fixed in a0eb0a801f.

The wrapper now defaults `INTEROP_CACHE_DIR` to `build/interop-cache` again, keeping the reusable verified baseline cache outside the wiped `build/interop-smoke` output directory. `INTEROP_CACHE_DIR` still overrides the cache location explicitly, and the README now documents the default cache path.

Validated:

- `bash -n tools/interop/run-hyphanet-interop-smoke.sh tools/interop/hyphanet-baseline.env`
- `INTEROP_SKIP_BUILD=1 bash -x tools/interop/run-hyphanet-interop-smoke.sh --self-test` confirmed `--download-cache-dir /work/cryptad/build/interop-cache`
- `INTEROP_SKIP_BUILD=1 INTEROP_CACHE_DIR=/tmp/cryptad-interop-cache-override bash -x tools/interop/run-hyphanet-interop-smoke.sh --self-test` confirmed explicit override is preserved
- `git diff --check`

File: tools/interop/interop_smoke.py [hunk 4]
```diff
@@ -269,19 +365,88 @@ def ensure_tcp_port_available(port: int) -> None:
...
-def ensure_deterministic_ports_available() -> None:
-    ensure_udp_port_available(DEFAULT_CRYPTAD_FNP_PORT)
-    ensure_tcp_port_available(DEFAULT_CRYPTAD_FCP_PORT)
-    ensure_udp_port_available(DEFAULT_HYPHANET_FNP_PORT)
-    ensure_tcp_port_available(DEFAULT_HYPHANET_FCP_PORT)
+def ensure_ports_available(ports: Ports) -> None:
+    ensure_udp_port_available(ports.cryptad_fnp)
+    ensure_tcp_port_available(ports.cryptad_fcp)
+    ensure_udp_port_available(ports.hyphanet_fnp)
+    ensure_tcp_port_available(ports.hyphanet_fcp)
```
[the remark this answers is not in the document]
Comment (leumor) [reply tools/interop/interop_smoke.py:391]: Fixed in 3d6b26c221.

`ensure_ports_available()` now rejects duplicate Cryptad/Hyphanet node bind assignments before probing sockets, specifically duplicate FNP UDP and duplicate FCP TCP ports. That turns an invalid override into an immediate configuration error instead of a later node startup bind failure. I also added self-test coverage for both duplicate cases.

Validated:

- `python3 -m py_compile tools/interop/interop_smoke.py`
- `python3 tools/interop/interop_smoke.py --self-test`
- `bash -n tools/interop/run-hyphanet-interop-smoke.sh tools/interop/hyphanet-baseline.env`
- `git diff --check`
- `INTEROP_SKIP_BUILD=1 INTEROP_CACHE_DIR=build/interop-cache tools/interop/run-hyphanet-interop-smoke.sh` passed in 134.795s

File: tools/interop/interop_smoke.py [hunk 5]
```diff
@@ -425,35 +597,75 @@ def download_asset() -> None:
...
+    asset_path = verified_download(url, expected_sha256, cache_dir)
+    if asset_path.suffix == ".jar":
+        return Baseline(version=version, asset_path=asset_path, kind="jar")
+    if asset_path.suffix == ".deb":
+        if shutil.which("dpkg-deb") is None:
+            raise InteropFailure("dpkg-deb is required to extract a Hyphanet .deb baseline")
+        extract_root = cache_dir / "hyphanet-root"
+        if extract_root.exists():
+            shutil.rmtree(extract_root)
+        subprocess.run(["dpkg-deb", "-x", str(asset_path), str(extract_root)], check=True)
```
[the remark this answers is not in the document]
Comment (leumor) [reply tools/interop/interop_smoke.py]: Fixed in 3d6b26c221.

The reusable cache now stores only the verified baseline package, while `.deb` extraction is staged under the per-run `build/interop-smoke/downloads/hyphanet-root` workspace. That avoids concurrent runs sharing the same `INTEROP_CACHE_DIR` from deleting or replacing each other’s extracted classpath while preserving baseline package reuse. The README now calls out the split between the reusable cache and per-run extracted contents.

Validated:

- `python3 -m py_compile tools/interop/interop_smoke.py`
- `python3 tools/interop/interop_smoke.py --self-test`
- `bash -n tools/interop/run-hyphanet-interop-smoke.sh tools/interop/hyphanet-baseline.env`
- `git diff --check`
- `INTEROP_SKIP_BUILD=1 INTEROP_CACHE_DIR=build/interop-cache tools/interop/run-hyphanet-interop-smoke.sh` passed in 134.795s
- Confirmed the live run staged the extracted root under `build/interop-smoke/downloads/hyphanet-root` while keeping the `.deb` asset in `build/interop-cache`.

File: tools/interop/interop_smoke.py [hunk 6]
```diff
@@ -396,24 +575,31 @@ def make_hyphanet_config(base_dir: Path, fnp_port: int, fcp_port: int) -> Path:
...
+        raise InteropFailure(f"Cannot determine baseline asset name from URL: {url}")
+    asset_path = cache_dir / parsed_name
     temp_asset_path = asset_path.with_name(asset_path.name + ".download")
 
     def download_asset() -> None:
-        if temp_asset_path.exists():
-            temp_asset_path.unlink()
-        urlretrieve(release_url, temp_asset_path)
+        temp_asset_path.unlink(missing_ok=True)
+        urlretrieve(url, temp_asset_path)
```
[the remark this answers is not in the document]
Comment (leumor) [reply tools/interop/interop_smoke.py]: Fixed in a522e36614.

`verified_download()` now reserves a unique temp file in the shared cache directory for each download attempt instead of using the fixed `<asset>.download` path. Each process cleans up only its own temp file on failure, and corrupt cached assets are removed with `missing_ok=True` before retrying so parallel repair/download attempts do not fail on a raced unlink. The final publish still uses `Path.replace()` into the cache asset path.

Validated:

- `python3 -m py_compile tools/interop/interop_smoke.py`
- `python3 tools/interop/interop_smoke.py --self-test`
- `bash -n tools/interop/run-hyphanet-interop-smoke.sh tools/interop/hyphanet-baseline.env`
- `git diff --check`
- local `file://` verified_download check covering fresh download and corrupt-cache replacement

I did not rerun the full live node smoke for this commit because the node launch/content flow was unchanged; the previous live smoke after the extraction/port fix passed, and this patch is isolated to download staging plus self-test coverage.
